File: assignments/clojure/anagram/anagram_test.clj
(ns anagram.test (:require [clojure.test :refer :all]))
(load-file "anagram.clj")

(deftest no-matches
  (is (= [] (anagram/anagrams-for "diaper" ["hello" "world" "zombies" "pants"]))))

(deftest detect-simple-anagram
  (is (= ["tan"] (anagram/anagrams-for "ant" ["tan" "stand" "at"]))))

(deftest does-not-confuse-different-duplicates
  (is (= [] (anagram/anagrams-for "galea" ["eagle"]))))

(deftest eliminate-anagram-subsets
  (is (= [] (anagram/anagrams-for "good" ["dog" "goody"]))))

(deftest detect-anagram
  (is (= ["inlets"] (anagram/anagrams-for "listen" ["enlists" "google" "inlets" "banana"]))))

(deftest multiple-anagrams
  (is (= ["gallery" "regally" "largely"]
         (anagram/anagrams-for "allergy" ["gallery" "ballerina" "regally" "clergy" "largely" "leading"]))))

(deftest case-insensitive-anagrams
  (is (= ["Carthorse"]
         (anagram/anagrams-for "Orchestra" ["cashregister" "Carthorse" "radishes"]))))

(deftest word-is-not-own-anagram
  (is (= [] (anagram/anagrams-for "banana" ["banana"]))))

(run-tests)

File: assignments/haskell/anagram/anagram_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import Anagram (anagramsFor)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly (runTestTT (TestList anagramTests))

anagramTests :: [Test]
anagramTests =
  [ testCase "no matches" $
    [] @=? anagramsFor "diaper" ["hello", "world", "zombies", "pants"]
  , testCase "detect simple anagram" $
    ["tan"] @=? anagramsFor "ant" ["tan", "stand", "at"]
  , testCase "does not confuse different duplicates" $
    [] @=? anagramsFor "galea" ["eagle"]
  , testCase "eliminate anagram subsets" $
    [] @=? anagramsFor "good" ["dog", "goody"]
  , testCase "detect anagram" $
    ["inlets"] @=? anagramsFor "listen" ["enlists", "google",
                                         "inlets", "banana"]
  , testCase "multiple anagrams" $
    ["gallery", "regally", "largely"] @=?
    anagramsFor "allergy" ["gallery", "ballerina", "regally", "clergy",
                           "largely", "leading"]
  , testCase "case insensitive anagrams" $
    ["Carthorse"] @=?
    anagramsFor "Orchestra" ["cashregister", "Carthorse", "radishes"]
  , testCase "does not detect a word as its own anagram" $
    [] @=? anagramsFor "banana" ["banana"]
  ]
